pkg/sfu/streamtracker.go

```diff
@@ -45,37 +50,69 @@ type StreamTracker struct {
 	lastSN uint16
 }
 
-func NewStreamTracker() *StreamTracker {
+func NewStreamTracker(samplesRequired uint32, cyclesRequired uint64, cycleDuration time.Duration) *StreamTracker {
 	s := &StreamTracker{
-		SamplesRequired: 5,
-		CyclesRequired:  60, // 30s of continuous stream
-		CycleDuration:   500 * time.Millisecond,
+		samplesRequired: samplesRequired,
+		cyclesRequired:  cyclesRequired,
+		cycleDuration:   cycleDuration,
+		status:          StreamStatusStopped,
 	}
-	s.status.set(int32(StreamStatusStopped))
 	return s
 }
 
+func (s *StreamTracker) OnStatusChanged(f func(status StreamStatus)) {
+	s.onStatusChanged = f
+}
+
 func (s *StreamTracker) Status() StreamStatus {
-	return StreamStatus(s.status.get())
+	s.statusMu.RLock()
+	defer s.statusMu.RUnlock()
+
+	return s.status
 }
 
-func (s *StreamTracker) setStatus(status StreamStatus) {
-	if status != s.Status() {
-		s.status.set(int32(status))
-		if s.OnStatusChanged != nil {
-			s.OnStatusChanged(status)
-		}
+func (s *StreamTracker) maybeSetActive() {
+	changed := false
+	s.statusMu.Lock()
+	if s.status != StreamStatusActive {
+		s.status = StreamStatusActive
+		changed = true
+	}
+	s.statusMu.Unlock()
+
+	if changed && s.onStatusChanged != nil {
+		s.onStatusChanged(StreamStatusActive)
 	}
 }
 
-func (s *StreamTracker) Start() {
+func (s *StreamTracker) maybeSetStopped() {
+	changed := false
+	s.statusMu.Lock()
+	if s.status != StreamStatusStopped {
+		s.status = StreamStatusStopped
+		changed = true
+	}
+	s.statusMu.Unlock()
+
+	if changed && s.onStatusChanged != nil {
+		s.onStatusChanged(StreamStatusStopped)
+	}
+}
+
+func (s *StreamTracker) init() {
+	s.initialized.set(true)
+	s.maybeSetActive()
+
 	if s.isRunning() {
```

Comment:
Moving start of detect worker after initialization so that they do not overlap. I still left in an empty `Start` method as a counterpart for `Stop`.